File: src/main/java/net/igneo/icv/enchantmentActions/enchantManagers/armor/chestplate/AbyssOmenManager.java
package net.igneo.icv.enchantmentActions.enchantManagers.armor.chestplate;

import net.igneo.icv.client.indicators.EnchantIndicator;
import net.igneo.icv.client.indicators.StasisCooldownIndicator;
import net.igneo.icv.enchantment.EnchantType;
import net.igneo.icv.enchantmentActions.EntityTracker;
import net.igneo.icv.enchantmentActions.enchantManagers.armor.ArmorEnchantManager;
import net.igneo.icv.entity.ICVEntity;
import net.igneo.icv.entity.ModEntities;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.phys.BlockHitResult;
import net.minecraft.world.phys.HitResult;
import net.minecraft.world.phys.Vec3;

public class AbyssOmenManager extends ArmorEnchantManager implements EntityTracker {
    private ICVEntity child = null;
    
    public AbyssOmenManager(Player player) {
        super(EnchantType.CHESTPLATE, 300, -10, false, player);
    }
    
    @Override
    public void onOffCoolDown(Player player) {
    
    }
    
    @Override
    public EnchantIndicator getIndicator() {
        return new StasisCooldownIndicator(this);
    }
    
    @Override
    public void activate() {
        if (!(player.level() instanceof ServerLevel level)) {
            return;
        }
        child = ModEntities.ABYSS_STONE.get().create(player.level());
        child.setOwner(player);
        
        HitResult hitResult = player.pick(5, 0f, false);
        
        BlockHitResult blockHitResult = (BlockHitResult) hitResult;
        Vec3 position = blockHitResult.getLocation();
        
        child.setPos(position);
        level.addFreshEntity(child);
        syncClientChild((ServerPlayer) player, child, this);
    }
    
    @Override
    public boolean shouldTickCooldown() {
        return child == null;
    }
    
    @Override
    public void tick() {
        super.tick();
        if (child != null && !child.isAlive()) {
            child = null;
        }
    }
    
    @Override
    public void onRemove() {
        if (this.child != null) {
            this.child.discard();
        }
    }
    
    @Override
    public boolean canUse() {
        return stableCheck() && child == null;
    }
    
    @Override
    public ICVEntity getChild() {
        return child;
    }
    
    @Override
    public void setChild(ICVEntity entity) {
        child = entity;
    }
}

File: src/main/java/net/igneo/icv/enchantmentActions/enchantManagers/armor/chestplate/AbyssOmenManager.kt
package net.igneo.icv.enchantmentActions.enchantManagers.armor.chestplate

import net.igneo.icv.client.indicators.EnchantIndicator
import net.igneo.icv.client.indicators.StasisCooldownIndicator
import net.igneo.icv.enchantment.EnchantType
import net.igneo.icv.enchantmentActions.EntityTracker
import net.igneo.icv.enchantmentActions.enchantManagers.armor.ArmorEnchantManager
import net.igneo.icv.entity.ABYSS_STONE
import net.igneo.icv.entity.ICVEntity
import net.minecraft.server.level.ServerLevel
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.entity.player.Player
import net.minecraft.world.phys.BlockHitResult

class AbyssOmenManager(player: Player?) :
    ArmorEnchantManager(EnchantType.CHESTPLATE, 300, -10, false, player), EntityTracker {
    override var child: ICVEntity? = null

    override fun onOffCoolDown(player: Player?) {
    }

    override val indicator: EnchantIndicator
        get() = StasisCooldownIndicator(this)

    override fun activate() {
        val level = player.level()
        if (player.level() !is ServerLevel) {
            return
        }
        child = ABYSS_STONE.get().create(player.level())
        child!!.owner = player

        val hitResult = player.pick(5.0, 0f, false)

        val blockHitResult = hitResult as BlockHitResult
        val position = blockHitResult.location

        child!!.setPos(position)
        level.addFreshEntity(child)
        syncClientChild(player as ServerPlayer, child, this)
    }

    override fun shouldTickCooldown(): Boolean {
        return child == null
    }

    override fun tick() {
        super.tick()
        if (child != null && !child!!.isAlive) {
            child = null
        }
    }

    override fun onRemove() {
        if (this.child != null) {
            child!!.discard()
        }
    }

    override fun canUse(): Boolean {
        return stableCheck() && child == null
    }
}